### ISSUE TITLE
Creating a BitwardenSecret loads all secrets in project into the destination secret

### ISSUE THREAD

**Issue by deefdragon** [bitwarden/sm-kubernetes#60]
I have created a BitwardenSecret with one item in the map, and the end result secret has literally every secret from my secret manager in it (the key being the secret's UUID), which is a non-trivial number of items.

I have certain tooling that expects exact key format for the secret, and having a whole bunch of UUID keys in the secret will be problimatic for that tooling, let alone the security implications of having all those extra secret values lying around.

[This line](https://github.com/bitwarden/sm-kubernetes/blob/10634a49d062dc9d97e10a0480b079be40dadb2f/internal/controller/bitwardensecret_controller.go#L260) is one problem line, adding ALL secrets tothe output secret. This would be fine on its own were it cleaned up, but [ApplySecretMap](https://github.com/bitwarden/sm-kubernetes/blob/10634a49d062dc9d97e10a0480b079be40dadb2f/internal/controller/bitwardensecret_controller.go#L282) run directly after it on the secret, only updates the k8s secret to have a key/value pair with a sane name instead of the UUID. 

A more sane approach would be to pass the secrets fetched directly to ApplySecretMap instead of setting it on the k8s secret in UpdateSecretValues.

**Comment by deefdragon** [comment]
After making the change mentioned, the tests failed. Fixing the tests, I have come to the conclusion that this is intended behavior? If so, I think this is not a good decision, and there should _at minimum_ be an option to disable this functionality. in a manner of least privilage, Secrets should be exposed as little as possible, and this was exposing EVERY secret to a given namespace.

**Comment by evanjarrett** [comment]
Can confirm this is an issue, or rather, I believe this is unexpected behavior.
If I'm filtering values, I would expect only the `bwSecretId` I specify to be exposed to the secret.
Why would I want all secrets to be exposed to an application if it doesn't need them?

A quick way to fix this would be to move [this delete call](https://github.com/bitwarden/sm-kubernetes/blob/10634a49d062dc9d97e10a0480b079be40dadb2f/internal/controller/bitwardensecret_controller.go#L291) one line down outside of the enclosing if, so that when a map is specified, only the secrets with `bwSecretId` maps remain.

The only other work around would be to create a new machine account for each kubes secret you want to create, which I think is going to make this project not very scalable... I would already be past the free tier, and for even modest deployments I could see this going past the 20 account limit.

**Comment by M4RC02U1F4A4** [comment]
Is there any news about a fix for this issue?

**Comment by deefdragon** [comment]
I've been running a custom variant of this that solves this problem and a few others, (and has proven stable), but it is brute force about the lot, so not ideal and I'm not ready to make a PR for it.

That and I need permission from my employer to contribute to open source projects due to certain contract shenanigans and I cant get _that_ until I'm off leave.

**Comment by evanjarrett** [comment]
I put together my own fork with this feature too, it just deletes out everything else in the list.
you can see the diff here (has some GHA changes so that I'm able to build it)
https://github.com/evanjarrett/sm-kubernetes/commit/8481fd3a3c1143ed9eeea6fd711b956f0fade8ea

This  does not seem like the original intent of this project, but I don't understand how you would want all your secrets stored in a single Secret object in Kubes.

**Comment by M4RC02U1F4A4** [comment]
I agree that having a single secret makes no sense, I also don't understand why this choice

If a user needs more than one secret, it seems that currently the only option is to create a machine account for each sync (then going to give permissions to individual values rather than to a project), which is not viable, since even with the Enterprise plan the limit is 50

**Comment by AndreaF17** [comment]
I agree as well, If we look at the security point of view....can any maintainer provide a feedback about it?

**Comment by Krikooo** [comment]
Any new on this issue ?

**Comment by eldarmus** [comment]
any news ?

**Comment by Tfinn92** [comment]
This is still the behavior we are seeing. The _nonsensical_ solution (without recoding the tool) is to have multiple machine accounts associated to specific secrets and projects, which imo, is insane.

**Comment by evanjarrett** [comment]
With the latest dev image, this is now working. I had to update the CRDs with the ones provided in this repo as opposed to the helm chart in order to get the new config for `onlyMappedSecrets: true`

CRDs were updated manually with:
`kubectl apply -f https://raw.githubusercontent.com/bitwarden/sm-kubernetes/refs/heads/main/config/crd/bases/k8s.bitwarden.com_bitwardensecrets.yaml`

**Comment by tangowithfoxtrot** [comment]
Thanks everyone. Closing this now that https://github.com/bitwarden/helm-charts/pull/286 and https://github.com/bitwarden/sm-kubernetes/pull/92 have been merged. These changes will be included in the next release.

**Comment by deefdragon** [comment]
Thank you @quisitive-crogers for doing the heavy lifting on this.

**Comment by hevel86** [comment]
IMO, this should be re-opened until the release is out, or at least add an updated version using the helm --devel flag.
